In [1]:
#import packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor

In [2]:
pwd

'C:\\Users\\seren\\Downloads\\Springboard\\capstone project 2\\Capstone project 2_'

In [3]:
df = pd.read_csv('California_Houses.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Median_House_Value        20640 non-null  float64
 1   Median_Income             20640 non-null  float64
 2   Median_Age                20640 non-null  int64  
 3   Tot_Rooms                 20640 non-null  int64  
 4   Tot_Bedrooms              20640 non-null  int64  
 5   Population                20640 non-null  int64  
 6   Households                20640 non-null  int64  
 7   Latitude                  20640 non-null  float64
 8   Longitude                 20640 non-null  float64
 9   Distance_to_coast         20640 non-null  float64
 10  Distance_to_LA            20640 non-null  float64
 11  Distance_to_SanDiego      20640 non-null  float64
 12  Distance_to_SanJose       20640 non-null  float64
 13  Distance_to_SanFrancisco  20640 non-null  float64
dtypes: flo

In [5]:
#X = df.drop('Median_House_Value', axis=1).values
#y = df['Median_House_Value'].values

In [6]:
X = df.iloc[:,1:]
X.head()

Median_Income  Median_Age  Tot_Rooms  Tot_Bedrooms  Population  Households  \
0         8.3252          41        880           129         322         126   
1         8.3014          21       7099          1106        2401        1138   
2         7.2574          52       1467           190         496         177   
3         5.6431          52       1274           235         558         219   
4         3.8462          52       1627           280         565         259   

   Latitude  Longitude  Distance_to_coast  Distance_to_LA  \
0     37.88    -122.23        9263.040773   556529.158342   
1     37.86    -122.22       10225.733072   554279.850069   
2     37.85    -122.24        8259.085109   554610.717069   
3     37.85    -122.25        7768.086571   555194.266086   
4     37.85    -122.25        7768.086571   555194.266086   

   Distance_to_SanDiego  Distance_to_SanJose  Distance_to_SanFrancisco  
0         735501.806984         67432.517001              21250.213767  
1         733236.884360         65049.908574              20880.600400  
2         733525.682937         64867.289833              18811.487450  
3         734095.290744         65287.138412              18031.047568  
4         734095.290744         65287.138412              18031.047568

In [7]:
y = df.iloc[:, 0]
y.head()

0    452600.0
1    358500.0
2    352100.0
3    341300.0
4    342200.0
Name: Median_House_Value, dtype: float64

In [8]:
#split data into train 80% and test 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [9]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(16512, 13) (16512,)
(4128, 13) (4128,)


In [10]:
#future scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train.iloc[:, :] = sc.fit_transform(X_train.iloc[:, :])
X_test.iloc[:, :] = sc.transform(X_test.iloc[:, :])

In [11]:
X_train.head()

Median_Income  Median_Age  Tot_Rooms  Tot_Bedrooms  Population  \
8944       -0.505293    1.138994  -0.673943     -0.553727   -0.917161   
17626       0.739144   -0.999198   0.590838      0.496533    0.413868   
18904      -0.622252    1.297378   0.809515      1.091287    0.604522   
6093        0.191586    0.347071  -0.265230     -0.419200   -0.314163   
16018       1.072242    1.851724  -0.231133     -0.617451   -0.435650   

       Households  Latitude  Longitude  Distance_to_coast  Distance_to_LA  \
8944    -0.651273 -0.757638   0.547282          -0.757010       -0.999673   
17626    0.615593  0.764052  -1.186561          -0.252163        0.887551   
18904    0.971980  1.172655  -1.351451          -0.701378        1.257168   
6093    -0.419751 -0.705976   0.847081           0.169856       -0.943623   
16018   -0.560225  0.989488  -1.451384          -0.702541        1.174304   

       Distance_to_SanDiego  Distance_to_SanJose  Distance_to_SanFrancisco  
8944              -0.722977             0.610578                  0.651176  
17626              0.931032            -1.558915                 -1.254856  
18904              1.250850            -1.177299                 -1.381123  
6093              -0.788567             0.737476                  0.761446  
16018              1.176752            -1.297581                 -1.521046

In [12]:
X_test.head()

Median_Income  Median_Age  Tot_Rooms  Tot_Bedrooms  Population  \
5856        0.424665    1.297378  -0.032914     -0.051018   -0.151886   
10653       0.508058   -1.870314   4.243335      4.617329    2.571360   
12512      -1.254147   -0.761622   0.332608      0.635781   -0.013551   
9329        0.047773   -0.999198   0.512187      0.708945    0.495450   
10084       0.619982   -1.711929  -0.014275     -0.338955   -0.838239   

       Households  Latitude  Longitude  Distance_to_coast  Distance_to_LA  \
5856     0.025083 -0.677797   0.617235          -0.326224       -1.020309   
10653    3.924530 -0.964288   0.917035          -0.694605       -0.796408   
12512    0.657214  1.374607  -0.941725           0.288220        1.247497   
9329     0.818499  1.092813  -1.451384          -0.782118        1.246280   
10084   -0.887997  1.750333  -0.312145           2.987674        1.398735   

       Distance_to_SanDiego  Distance_to_SanJose  Distance_to_SanFrancisco  
5856              -0.697874             0.582415                  0.627078  
10653             -0.998735             0.985084                  0.976600  
12512              1.241144            -0.960136                 -1.058891  
9329               1.239920            -1.214913                 -1.465498  
10084              1.347634            -0.370188                 -0.498250

In [13]:
rf = RandomForestRegressor().fit(X_train, y_train)
rf.score(X_test, y_test)

0.8334932703888838

In [14]:
lr = LinearRegression().fit(X_train, y_train)
lr.score(X_test, y_test)

0.6378161111012643

In [15]:
from sklearn.linear_model import Lasso
la = Lasso().fit(X_train, y_train)
la.score(X_test, y_test)

C:\Users\seren\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.341e+13, tolerance: 2.201e+10
  model = cd_fast.enet_coordinate_descent(


0.637787745866392

In [16]:
from xgboost import XGBRegressor

regressor = XGBRegressor()
regressor.fit(X_train, y_train)

XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
             colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
             early_stopping_rounds=None, enable_categorical=False,
             eval_metric=None, feature_types=None, gamma=0, gpu_id=-1,
             grow_policy='depthwise', importance_type=None,
             interaction_constraints='', learning_rate=0.300000012, max_bin=256,
             max_cat_threshold=64, max_cat_to_onehot=4, max_delta_step=0,
             max_depth=6, max_leaves=0, min_child_weight=1, missing=nan,
             monotone_constraints='()', n_estimators=100, n_jobs=0,
             num_parallel_tree=1, predictor='auto', random_state=0, ...)

In [18]:
regressor.score(X_test, y_test)

0.8342009020324609

#R^2 value on each model

In [22]:
#R^2 value on xgboost
from sklearn.metrics import r2_score 
y_pred_xgboost = regressor.predict(X_test)
r2_score(y_test, y_pred_xgboost)

0.8342009020324609

it looks like the xgboost with R^2 0.834 and random forest model with R^2 0.833 perform better than other models